liquid-compiler/src/parser.rs

```diff
@@ -3,477 +3,831 @@
+        )
+    }
 }
 
-// a tag can be either a single-element tag or a block, which can contain other
-// elements and is delimited by a closing tag named {{end +
-// the_name_of_the_tag}}. Tags do not get rendered, but blocks may contain
-// renderable expressions
-fn parse_tag(
-    iter: &mut Iter<Element>,
-    tokens: &[Token],
-    options: &LiquidOptions,
-) -> Result<Box<Renderable>> {
-    let tag = &tokens[0];
-    match *tag {
-        // is a tag
-        Token::Identifier(ref x) if options.tags.contains_key(x.as_str()) => {
-            options.tags[x.as_str()].parse(x, &tokens[1..], options)
+/// An element that is raw text.
+pub struct Raw<'a> {
+    text: &'a str,
+}
+impl<'a> From<Pair<'a>> for Raw<'a> {
+    fn from(element: Pair<'a>) -> Self {
+        if element.as_rule() != Rule::Raw {
+            panic!("Only rule Raw can be converted to Raw.");
         }
+        Raw {
+            text: element.as_str(),
+        }
+    }
+}
+impl<'a> Into<&'a str> for Raw<'a> {
+    fn into(self) -> &'a str {
+        self.to_str()
+    }
+}
+impl<'a> Raw<'a> {
+    /// Turns the text into a Renderable.
+    pub fn to_renderable(self) -> Box<Renderable> {
+        Box::new(Text::new(self.to_str()))
+    }
 
-        // is a block
-        Token::Identifier(ref x) if options.blocks.contains_key(x.as_str()) => {
-            // Collect all the inner elements of this block until we find a
-            // matching "end<blockname>" tag. Note that there may be nested blocks
-            // of the same type (and hence have the same closing delimiter) *inside*
-            // the body of the block, which would premauturely stop the element
-            // collection early if we did a nesting-unaware search for the
-            // closing tag.
-            //
-            // The whole nesting count machinery below is to ensure we only stop
-            // collecting elements when we have an un-nested closing tag.
-
-            let end_tag = Token::Identifier(format!("end{}", x));
-            let mut children = vec![];
-            let mut nesting_depth = 0;
-            for t in iter {
-                if let Element::Tag(ref tokens, _) = *t {
-                    match tokens[0] {
-                        ref n if n == tag => {
-                            nesting_depth += 1;
-                        }
-                        ref n if n == &end_tag && nesting_depth > 0 => {
-                            nesting_depth -= 1;
-                        }
-                        ref n if n == &end_tag && nesting_depth == 0 => break,
-                        _ => {}
-                    }
-                };
-                children.push(t.clone())
-            }
-            options.blocks[x.as_str()].parse(x, &tokens[1..], &children, options)
+    /// Converts the text into a str.
+    pub fn to_str(self) -> &'a str {
+        self.as_str()
+    }
+
+    /// Returns the text as a str.
+    pub fn as_str(&self) -> &'a str {
```

Comment:
I remember it making sense at the time I wrote it, but right now I honestly can't see why this would be needed. Maybe the code was changed in a way that it stopped being needed or I was thinking wrong.